plugin/trino-loki/src/main/java/io/trino/plugin/loki/LokiRecordCursor.java

```diff
@@ -0,0 +1,159 @@
+    {
+        this.columnHandles = ImmutableList.copyOf(columnHandles);
+        this.resultIterator = queryResult;
+
+        fieldToColumnIndex = new int[columnHandles.size()];
+        for (int i = 0; i < columnHandles.size(); i++) {
+            LokiColumnHandle columnHandle = columnHandles.get(i);
+            fieldToColumnIndex[i] = columnHandle.ordinalPosition();
+        }
+    }
+
+    @Override
+    public boolean advanceNextPosition()
+    {
+        return this.resultIterator.advanceNextPosition();
+    }
+
+    @Override
+    public long getCompletedBytes()
+    {
+        return 0;
+    }
+
+    @Override
+    public long getReadTimeNanos()
+    {
+        return 0;
+    }
+
+    @Override
+    public Type getType(int field)
+    {
+        checkArgument(field < columnHandles.size(), "Invalid field index");
+        return columnHandles.get(field).type();
+    }
+
+    Object getFieldValue(int field)
+    {
+        // TODO: checkState(current != null, "Cursor has not been advanced yet");
+
+        int columnIndex = fieldToColumnIndex[field];
+        return switch (columnIndex) {
+            case 0 -> getSqlMapFromLabels(this.columnHandles.get(columnIndex).type(), this.resultIterator.getLabels());
+            case 1 -> this.resultIterator.getTimestamp();
+            case 2 -> this.resultIterator.getValue();
+            default -> null;
+        };
+    }
+
+    @Override
+    public boolean getBoolean(int field)
+    {
+        return false;
+    }
+
+    @Override
+    public long getLong(int field)
+    {
+        checkFieldType(field, LokiMetadata.TIMESTAMP_COLUMN_TYPE);
+        return (long) requireNonNull(getFieldValue(field));
+    }
+
+    @Override
+    public double getDouble(int field)
+    {
+        checkFieldType(field, DOUBLE);
+        return (double) requireNonNull(getFieldValue(field));
+    }
+
+    @Override
+    public Slice getSlice(int field)
+    {
+        checkFieldType(field, createUnboundedVarcharType());
+        return Slices.utf8Slice((String) requireNonNull(getFieldValue(field)));
+    }
+
+    @Override
+    public Object getObject(int field)
+    {
+        return getFieldValue(field);
+    }
+
+    @Override
+    public boolean isNull(int field)
+    {
+        checkArgument(field < columnHandles.size(), "Invalid field index");
+        return getFieldValue(field) == null;
+    }
+
+    @Override
+    public void close()
+    {
+    }
+
+    private void checkFieldType(int field, Type expected)
+    {
+        Type actual = getType(field);
+        checkArgument(actual.equals(expected), "Expected field %s to be type %s but is %s", field, expected, actual);
+    }
+
+    // Copy from Loki to handle map<string,string>
+    static SqlMap getSqlMapFromLabels(Type type, Map<String, String> labels)
+    {
+        // on functions like COUNT() the Type won't be a MapType
+        if (!(type instanceof MapType mapType)) {
```

Comment:
Is it possible to avoid calling this method with an invalid type? I don't understand the comment about `COUNT()`.

Reply:
This is straight from the Prometheus connector. I actually hope to go to the page provider and avoid this class altogether...I removed it. It was really not required :slightly_smiling_face: